deploy/runtools/run_farm_deploy_managers.py

```diff
@@ -656,6 +661,37 @@ def clear_fpgas(self) -> None:
                 with StreamLogger('stdout'), StreamLogger('stderr'):
                     run(f"xbutil reset -d {card_bdf} --force")
 
+    def localize_xclbin(self, slotno: int) -> None:
+        """ download xclbin URI to remote node. """
+        assert slotno < len(self.parent_node.sim_slots)
+        serv = self.parent_node.sim_slots[slotno]
+        hwcfg = serv.get_resolved_server_hardware_config()
+        if re.match(_RFC_3986_PATTERN, hwcfg.xclbin):
+            remote_home_dir = self.parent_node.get_sim_dir()
+            remote_sim_dir = """{}/sim_slot_{}/""".format(remote_home_dir, slotno)
+            hwcfg.local_xclbin = './local.xclbin'
+
+            def get_xclbin(xclbin_uri:str, xclbin_lpath:PathLike) -> None:
+                # TODO instead of blowing up if the file exists, consider using fsspec
+                # filecache https://filesystem-spec.readthedocs.io/en/latest/features.html#caching-files-locally
+                # so that multiple slots using the same xclbin only grab it once and
+                # we only download it if it has changed at the source.
+                # HOWEVER, 'filecache' isn't thread/process safe and I'm not sure whether
+                # this runs in @parallel for fabric
+                lpath = Path(xclbin_lpath)
+                assert not lpath.exists, f"{lpath.resolve(strict=False)} already exists, refusing to overwrite"
+                rootLogger.debug("Downloading '%s' to '%s'", xclbin_uri, )
+                fs, rpath = url_to_fs(xclbin_uri)
+                fs.get_file(rpath, fspath(lpath)) # fsspec deals in strings, not PathLike
```

Comment:
Whats the debug output if you can't obtain a file? Does this get caught by the `StreamLogger` (I forget how `StreamLogger` works exactly)

Reply:
Good question.  I think it throws an exception and I would suspect that the debug output quality varies based on which filesystem implementation is being used.  When I tried to `fs.ls()` a non-existent file (outside of FireSim and fabric), a stacktrace looked like :
```
FileNotFoundError                         Traceback (most recent call last)
/home/centos/orca_pr_220621/fsspec.ipynb Cell 13 in <cell line: 1>()
----> [1](vscode-notebook-cell://ssh-remote%2Bfs5/home/centos/orca_pr_220621/fsspec.ipynb#ch0000012vscode-remote?line=0) fs[0].ls(fs[1])

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/fsspec/implementations/sftp.py:115, in SFTPFileSystem.ls(self, path, detail)
    113 def ls(self, path, detail=False):
    114     logger.debug("Listing folder %s" % path)
--> 115     stats = [self._decode_stat(stat, path) for stat in self.ftp.listdir_iter(path)]
    116     if detail:
    117         return stats

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/fsspec/implementations/sftp.py:115, in <listcomp>(.0)
    113 def ls(self, path, detail=False):
    114     logger.debug("Listing folder %s" % path)
--> 115     stats = [self._decode_stat(stat, path) for stat in self.ftp.listdir_iter(path)]
    116     if detail:
    117         return stats

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/paramiko/sftp_client.py:278, in SFTPClient.listdir_iter(self, path, read_aheads)
    276 path = self._adjust_cwd(path)
    277 self._log(DEBUG, "listdir({!r})".format(path))
--> 278 t, msg = self._request(CMD_OPENDIR, path)
    280 if t != CMD_HANDLE:
    281     raise SFTPError("Expected handle")

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/paramiko/sftp_client.py:822, in SFTPClient._request(self, t, *arg)
    820 def _request(self, t, *arg):
    821     num = self._async_request(type(None), t, *arg)
--> 822     return self._read_response(num)

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/paramiko/sftp_client.py:874, in SFTPClient._read_response(self, waitfor)
    871 if num == waitfor:
    872     # synchronous
    873     if t == CMD_STATUS:
--> 874         self._convert_status(msg)
    875     return t, msg
    877 # can not rewrite this to deal with E721, either as a None check
    878 # nor as not an instance of None or NoneType

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/paramiko/sftp_client.py:903, in SFTPClient._convert_status(self, msg)
    900     raise EOFError(text)
    901 elif code == SFTP_NO_SUCH_FILE:
    902     # clever idea from john a. meinel: map the error codes to errno
--> 903     raise IOError(errno.ENOENT, text)
    904 elif code == SFTP_PERMISSION_DENIED:
    905     raise IOError(errno.EACCES, text)

FileNotFoundError: [Errno 2] No such file
```

I suspect that we will want to put this into  a `try/except` block that makes the debug output nicer than that big stacktrace.

Reply:
I think this debug output is fine. I rather it be very verbose and descriptive without us having to do extra work to catch an exception.

Reply:
cool.  I won't worry about adding nicer error messaging for now then.